app/lib/brexit_checker/results_audiences.rb

```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
     def populate_citizen_groups(audience_actions, selected_criteria)
       return [] if audience_actions.blank? || selected_criteria.blank?
 
-      BrexitChecker::Group.load_all.each_with_object([]) do |group, result|
-        selected_actions = group.actions & audience_actions
-        unless selected_actions.empty?
-          result << {
+      audience_actions.each_with_object([]) do |action, grouped_actions|
+        next if action.grouping_criteria.empty?
+
+        group_keys =
+          multiple_grouping_criteria?(action) ? selected_group(action, selected_criteria) : action.grouping_criteria
+        group_keys.each do |key|
+          group = all_groups.select { |g| g.key == key }.first
+          next if grouped_actions.select { |actions| actions[:group] == group }.any?
+
+          grouped_actions << {
             group: group,
-            actions: selected_actions,
-            criteria: selected_actions.flat_map(&:all_criteria).uniq & selected_criteria,
+            actions: actions_for_group(audience_actions, group),
+            criteria: criteria_for_group(audience_actions, group, selected_criteria),
           }
+          grouped_actions.sort_by! { |actions| [(actions[:group].priority * -1), actions[:group].key] }
         end
-        result
       end
     end
+
+    def selected_group(action, selected_criteria)
```

Comment:
Is this named correctly? It seems to be more `grouping_criteria` rather than `group`, or something?

Reply:
I've renamed `selected_grouping_criteria`, as it's the grouping criteria which have a key in common with the user's selected criteria, rather than all of that action's `grouping_criteria`.